### PR TITLE
cpu/gd32v: add periph_i2c support

### PR DESCRIPTION
**Pull request by gschorcht**
### Contribution description

This PR provides the `periph_i2c` support and is one of a bunch of PRs that complete the peripheral drivers for GD32VF103.

The driver is a modified version of the driver for STM32F1 with some changes that were necessary to get it working on GD32V. As  for STM32F1, the driver is using polling instead of interrupts for now. It will be implemented interrupt-driven later.

### Testing procedure

`tests/periph_i2c` as well as a test with any I2C sensor should work. The driver was tested with `tests/driver_l3gxxxx` and `tests/driver_bmp180`.

### CODE REVIEW

File: boards/common/gd32v/include/cfg_i2c_default.h [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+#define CFG_I2C_DEFAULT_H
+
+#include "periph_cpu.h"
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @name    I2C configuration
```
Comment (benpicco) [inline boards/common/gd32v/include/cfg_i2c_default.h:29]:
```suggestion
 * @brief    I2C configuration
```
should make doxygen happy

File: boards/common/gd32v/include/cfg_i2c_default.h [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+        .dev            = I2C1,
+        .speed          = I2C_SPEED_NORMAL,
+        .scl_pin        = GPIO_PIN(PORT_B, 10),
+        .sda_pin        = GPIO_PIN(PORT_B, 11),
+        .rcu_mask       = RCU_APB1EN_I2C1EN_Msk,
+        .irqn           = I2C1_EV_IRQn,
+    }
+};
+
+#define I2C_NUMOF   ARRAY_SIZE(i2c_config)
```
Comment (benpicco) [inline boards/common/gd32v/include/cfg_i2c_default.h:53]:
```suggestion
#define I2C_NUMOF   ARRAY_SIZE(i2c_config)    /**< number of I2C channels */
```

File: cpu/gd32v/include/vendor/gd32vf103_periph.h [hunk 3]
```diff
@@ -11549,8 +11549,8 @@ typedef struct {                                /*!< (@ 0x40002C00) WWDGT Struct
 //#define GPIOC_BASE                  0x40011000UL
 //#define GPIOD_BASE                  0x40011400UL
 //#define GPIOE_BASE                  0x40011800UL
-//#define I2C0_BASE                   0x40005400UL
-//#define I2C1_BASE                   0x40005800UL
+#define I2C0_BASE                   0x40005400UL
+#define I2C1_BASE                   0x40005800UL
```
Comment (benpicco) [inline cpu/gd32v/include/vendor/gd32vf103_periph.h:11553]: why were they commented out in the first place :exploding_head:

File: cpu/gd32v/periph/i2c.c [hunk 4]
```diff
@@ -0,0 +1,476 @@
...
+    }
+    if (state & I2C_STAT0_SMBTO_Msk) {
+        DEBUG("TIMEOUT\n");
+    }
+    if (state & I2C_STAT0_SMBALT_Msk) {
+        DEBUG("SMBALERT\n");
+    }
+    core_panic(PANIC_GENERAL_ERROR, "I2C FAULT");
+}
+
```
Comment (benpicco) [inline cpu/gd32v/periph/i2c.c:476]:
```suggestion
```

File: cpu/gd32v/periph/i2c.c [hunk 5]
```diff
@@ -0,0 +1,476 @@
...
+void i2c_init(i2c_t dev)
+{
+    assert(dev < I2C_NUMOF);
+
+    mutex_init(&locks[dev]);
+
+    assert(i2c_config[dev].dev != NULL);
+
+    /* Configure pins in idle state as open drain outpus to keep the bus lines
```
Comment (benpicco) [inline cpu/gd32v/periph/i2c.c:92]:
```suggestion
    /* Configure pins in idle state as open drain outputs to keep the bus lines
```

File: cpu/gd32v/periph/i2c.c [hunk 6]
```diff
@@ -0,0 +1,476 @@
...
+    if (state & I2C_STAT0_PECERR_Msk) {
+        DEBUG("PECERR\n");
+    }
+    if (state & I2C_STAT0_SMBTO_Msk) {
+        DEBUG("TIMEOUT\n");
+    }
+    if (state & I2C_STAT0_SMBALT_Msk) {
+        DEBUG("SMBALERT\n");
+    }
+    core_panic(PANIC_GENERAL_ERROR, "I2C FAULT");
```
Comment (benpicco) [inline cpu/gd32v/periph/i2c.c:474]: I don't get this interrupt handler.
It's just printing the ISR flags, then panics?

Comment (gschorcht) [reply cpu/gd32v/periph/i2c.c:474]: Yes, it really doesn't make sense. For example, the loss of arbitration is a normal situation. But this is the implementation of STM32F1. In an intermediate version I had already removed this `core_panic`, but I forgot to remove it in the final version.

File: boards/common/gd32v/include/cfg_i2c_default.h (hunk 1 above)
Comment (gschorcht) [reply boards/common/gd32v/include/cfg_i2c_default.h:29]: Yes, but this is a group definition by intention. It makes probably more sense to include it in `generic_exclude_pattern`.

Comment (gschorcht) [reply boards/common/gd32v/include/cfg_i2c_default.h:29]: This is the way we use for all boards.

File: cpu/gd32v/include/vendor/gd32vf103_periph.h (hunk 3 above)
Comment (gschorcht) [reply cpu/gd32v/include/vendor/gd32vf103_periph.h:11553]: I don't know, we have to ask @bergzand. The file seems to be generated with SDVconv from the gs32vf103.svd file and these components were probably not selected?

**Comment by benpicco** [comment]
When you rebase, just squash directly.

**Comment by gschorcht** [comment]
> Do you want to keep the `_irq_handler()`? As I see it the driver is not interrupt based

Yeah, the `_irq_handler` makes absolutely no sense in that way. I removed it completely.

>  (it would of course be great if it were, but the polling driver won't need an interrupt handler).

I have it on my ToDo list and will continue my work as soon as I have time. At the moment the polling driver is not better or worse than the one for STM32F1.

File: cpu/gd32v/periph/i2c.c [hunk 7]
```diff
@@ -0,0 +1,428 @@
...
+    _init(dev);
+
+    periph_clk_dis(APB1, i2c_config[dev].rcu_mask);
+}
+
+static void _init_pins(i2c_t dev)
+{
+    /* This is needed in case the remapped pins are used */
+    if (i2c_config[dev].scl_pin == GPIO_PIN(PORT_B, 8) ||
+        i2c_config[dev].sda_pin == GPIO_PIN(PORT_B, 9)) {
```
Comment (benpicco) [inline cpu/gd32v/periph/i2c.c:106]: Is this also true for GD32V? IIRC this was about SWD

File: cpu/gd32v/periph/i2c.c [hunk 8]
```diff
@@ -0,0 +1,428 @@
...
+        case I2C_SPEED_NORMAL:
+            _init_clk(i2c, KHZ(100));
+            break;
+        case I2C_SPEED_FAST:
+            _init_clk(i2c, KHZ(400));
+            break;
+        case I2C_SPEED_FAST_PLUS:
+            _init_clk(i2c, MHZ(1));
+            break;
+    }
```
Comment (benpicco) [inline cpu/gd32v/periph/i2c.c:155]:
```suggestion
    _init_clk(i2c, i2c_config[dev].speed));
```

File: cpu/gd32v/periph/i2c.c (hunk 7 above)
Comment (gschorcht) [reply cpu/gd32v/periph/i2c.c:106]: > Is this also true for GD32V?

Yes, I2C0_SCL and I2C0_SDA signals could be remapped to PB8 and PB9 if a board specific configuration is used.

File: cpu/gd32v/include/periph_cpu.h [hunk 9]
```diff
@@ -243,8 +243,10 @@ typedef struct {
  */
 #define HAVE_I2C_SPEED_T
 typedef enum {
-    I2C_SPEED_NORMAL,       /**< normal mode:  ~100kbit/s */
-    I2C_SPEED_FAST,         /**< fast mode:    ~400kbit/s */
+    I2C_SPEED_LOW,          /**< low speed mode: ~10kit/s */
+    I2C_SPEED_NORMAL,       /**< normal mode:    ~100kbit/s */
+    I2C_SPEED_FAST,         /**< fast mode:      ~400kbit/s */
+    I2C_SPEED_FAST_PLUS,    /**< fast plus mode: ~1Mbit/s */
```
Comment (benpicco) [inline cpu/gd32v/include/periph_cpu.h:249]:
```suggestion
    I2C_SPEED_LOW       = KHZ(10),   /**< low speed mode: ~10kit/s */
    I2C_SPEED_NORMAL    = KHZ(100),  /**< normal mode:    ~100kbit/s */
    I2C_SPEED_FAST      = KHZ(400),  /**< fast mode:      ~400kbit/s */
    I2C_SPEED_FAST_PLUS = MHZ(1),    /**< fast plus mode: ~1Mbit/s */
```

File: cpu/gd32v/periph/i2c.c (hunk 7 above)
Comment (gschorcht) [reply cpu/gd32v/periph/i2c.c:106]: > IIRC this was about SWD

IMHO this has nothing to do with SWD. Some of the signals can be remapped to other GPIOs. This is true for GD32V as well as for STM32F1.

File: cpu/gd32v/periph/i2c.c (hunk 8 above)
Comment (gschorcht) [reply cpu/gd32v/periph/i2c.c:155]: :+1: cool solution.

**Review by benpicco — APPROVED** [review]
Please squash

**Comment by benpicco** [comment]
bors merge

**Pull request merged**